src/ansi.rs

```diff
@@ -0,0 +1,1849 @@
+                14 => handler.text_area_size_pixels(writer),
+                18 => handler.text_area_size_chars(writer),
+                22 => handler.push_title(),
+                23 => handler.pop_title(),
+                _ => unhandled!(),
+            },
+            ('u', None) => handler.restore_cursor_position(),
+            ('X', None) => handler.erase_chars(next_param_or(1) as usize),
+            ('Z', None) => handler.move_backward_tabs(next_param_or(1) as i64),
+            _ => unhandled!(),
+        }
+    }
+
+    #[inline]
+    fn esc_dispatch(&mut self, intermediates: &[u8], _ignore: bool, byte: u8) {
+        macro_rules! unhandled {
+            () => {{
+                debug!(
+                    "[unhandled] esc_dispatch ints={:?}, byte={:?} ({:02x})",
+                    intermediates, byte as char, byte
+                );
+            }};
+        }
+
+        macro_rules! configure_charset {
+            ($charset:path, $intermediate:expr) => {{
+                let index: CharsetIndex = match $intermediate {
+                    Some(b'(') => CharsetIndex::G0,
+                    Some(b')') => CharsetIndex::G1,
+                    Some(b'*') => CharsetIndex::G2,
+                    Some(b'+') => CharsetIndex::G3,
+                    _ => {
+                        unhandled!();
+                        return;
+                    },
+                };
+                self.handler.configure_charset(index, $charset)
+            }};
+        }
+
+        match (byte, intermediates.get(0)) {
+            (b'B', intermediate) => configure_charset!(StandardCharset::Ascii, intermediate),
+            (b'D', None) => self.handler.linefeed(),
+            (b'E', None) => {
+                self.handler.linefeed();
+                self.handler.carriage_return();
+            },
+            (b'H', None) => self.handler.set_horizontal_tabstop(),
+            (b'M', None) => self.handler.reverse_index(),
+            (b'Z', None) => self.handler.identify_terminal(self.writer, None),
+            (b'c', None) => self.handler.reset_state(),
+            (b'0', intermediate) => {
+                configure_charset!(StandardCharset::SpecialCharacterAndLineDrawing, intermediate)
+            },
+            (b'7', None) => self.handler.save_cursor_position(),
+            (b'8', Some(b'#')) => self.handler.decaln(),
+            (b'8', None) => self.handler.restore_cursor_position(),
+            (b'=', None) => self.handler.set_keypad_application_mode(),
+            (b'>', None) => self.handler.unset_keypad_application_mode(),
+            // String terminator, do nothing (parser handles as string terminator).
+            (b'\\', None) => (),
+            _ => unhandled!(),
+        }
+    }
+}
+
+fn attrs_from_sgr_parameters(
+    params: &mut ParamsIter<'_>,
+) -> impl IntoIterator<Item = Option<Attr>> {
+    #[cfg(not(feature = "alloc"))]
+    let mut attrs = ArrayVec::<[_; crate::params::MAX_PARAMS]>::new();
```

Comment:
No fully qualified paths please. Segments should be imported as appropriate at the beginning of the file.